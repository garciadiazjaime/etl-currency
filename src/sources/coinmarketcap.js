
import config from '../config';
import extract from '../util/extract';
import transform from '../transformer/coinmarketcap';
import load from '../util/load';

const props = {
  isProduction: config.get('env') === 'production',
  file: './stubs/coinmarketcap.json',
  fixerUrl: 'https://api.coinmarketcap.com/v2/ticker/',
  apiUrl: config.get('api.url'),
  currencies: ['BTC', 'LTC', 'ETH', 'ZEC', 'DASH', 'XRP', 'XMR'],
};


const main = () => extract(props)
  .then(response => transform(props, response))
  .then(currencies => load(props, currencies))
  .catch(console.log);


export default main;

