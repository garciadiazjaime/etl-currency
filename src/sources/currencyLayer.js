 import config from '../config'
 import extract from '../util/extract.js'
 import load from '../util/load'

 const props = {
   isProduction: config.get('env') === 'production',
   file: './stubs/currencyLayer.json',
   sourceUrl:`http://apilayer.net/api/live?access_key=${config.get('currencyLayer.token')}0&currencies=USD,AUD,CAD,PLN,MXN&format=1`,
   apiUrl: config.get('api.url'),
   currencies: ['USDPLN', 'USDCAD', 'USDMXN']
 }

 const transform = (props, response) => {
   const data = JSON.parse(response)
   if (!data || !data.success) {
     throw('Fixer return invalid response')
   }

   const { quotes } = data
   if (!Object.keys(quotes)) {
     throw('Fixer return invalid response')
   }

   const wantedRates = Object.keys(quotes).reduce((accumulator, currency) => {
     if (props.currencies.includes(currency)) {
       accumulator.push({
         [currency]: quotes[currency]
       })
     }
     return accumulator
   }, [])

   return wantedRates
 }

 const main = async () => {
   return extract(props)
     .then(response => transform(props, response))
     .then(currencies => load(props, currencies))
     .catch(console.log)
 }

 export default main

  
