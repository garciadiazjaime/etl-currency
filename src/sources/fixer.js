
import config from '../config'
import extract from '../util/extract.js'
import load from '../util/load'

const props = {
  isProduction: config.get('env') === 'production',
  file: './stubs/fixer.json',
  fixerUrl: `http://data.fixer.io/api/latest?access_key=${config.get('fixer.token')}&format=1`,
  apiUrl: config.get('api.url'),
  currencies: ['EUR', 'USD', 'MXN']
}

const transform = (props, response) => {
  const data = JSON.parse(response)
  if (!data || !data.success) {
    throw('Fixer return invalid response')
  }

  const { rates } = data
  if (!Object.keys(rates)) {
    throw('Fixer return invalid response')
  }

  const wantedRates = Object.keys(rates).reduce((accumulator, currency) => {
    if (props.currencies.includes(currency)) {
      accumulator.push({
        [currency]: rates[currency]
      })
    }
    return accumulator
  }, [])

  return wantedRates
}

const main = async () => {
  return extract(props)
    .then(response => transform(props, response))
    .then(currencies => load(props, currencies))
    .catch(console.log)
}


export default main


