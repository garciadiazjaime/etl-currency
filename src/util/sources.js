
import config from '../config';
import transformer from '../transformer';
import extract from './extract';
import load from './load';

const run = props => extract(props)
  .then(response => transformer[props.source](props, response))
  .then(currencies => load(props, currencies))
  .catch(console.log);

async function main() {
  const sources = config.get('sources');

  const promises = Object.keys(sources).reduce((accumulator, key) => {
    const source = sources[key];
    if (source.enable) {
      const props = {
        source: key,
        isProduction: config.get('env') === 'production',
        file: source.stubFile,
        sourceUrl: source.sourceUrl.replace('[token]', source.token),
        apiUrl: source.apiUrl,
        currencies: source.currencies,
      };
      accumulator.push(run(props));
    }
    return accumulator;
  }, []);

  return Promise.all(promises);
}

export default main;
