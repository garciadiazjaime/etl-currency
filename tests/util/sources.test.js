import sources from '../../src/util/sources';

describe('sources', () => {
  it('returns expected values', async () => {
    const response = await sources();
    expect(response).toEqual([
      [
        {
          CHF: 1.197226,
        },
        {
          CNY: 7.652477,
        },
        {
          EUR: 1,
        },
        {
          GBP: 0.877709,
        },
        {
          KWD: 0.36339,
        },
        {
          KYD: 0.991168,
        },
        {
          MXN: 22.623886,
        },
        {
          USD: 1.208463,
        },
      ],
      [
        {
          BTC: 9870.68,
        }, {
          LTC: 179.561,
        }, {
          XRP: 0.912325,
        }, {
          DASH: 505.126,
        }, {
          XMR: 243.154,
        }, {
          ETH: 811.324,
        }, {
          ZEC: 306.783,
        },
      ],
      [
        {
          USDCAD: 1.28465,
        }, {
          USDPLN: 3.567027,
        }, {
          USDMXN: 19.036598,
        },
      ],
    ]);
  });
});
