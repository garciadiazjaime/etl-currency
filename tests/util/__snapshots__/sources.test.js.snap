// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sources returns expected values 1`] = `
Array [
  Array [
    Object {
      "currency": "CHF",
      "rate": 0.9907,
      "type": false,
    },
    Object {
      "currency": "CNY",
      "rate": 6.3324,
      "type": false,
    },
    Object {
      "currency": "EUR",
      "rate": 0.8274,
      "type": false,
    },
    Object {
      "currency": "GBP",
      "rate": 0.7263,
      "type": false,
    },
    Object {
      "currency": "KWD",
      "rate": 0.3007,
      "type": false,
    },
    Object {
      "currency": "KYD",
      "rate": 0.8201,
      "type": false,
    },
    Object {
      "currency": "MXN",
      "rate": 18.7212,
      "type": false,
    },
    Object {
      "currency": "USD",
      "rate": 1,
      "type": false,
    },
  ],
]
`;
